### PR TITLE
*: replace `quick-error` with `thiserror` partially

### PR DESCRIPTION
**Pull request by koushiro**
### What problem does this PR solve?

Problem Summary: part of #8698

### What is changed and how it works?

What's Changed: replace `quick-error` with `thiserror` in some crates

### Check List 

Tests 

- No code

Side effects

- No

### Release note 

No

### CODE REVIEW

File: components/engine_traits/src/errors.rs [hunk 1]
```diff
@@ -1,55 +1,41 @@
...
+    Codec(#[from] tikv_util::codec::Error),
+    #[error("The entries of region is unavailable")]
+    EntriesUnavailable,
+    #[error("The entries of region is compacted")]
+    EntriesCompacted,
+}
+
+impl From<String> for Error {
+    fn from(err: String) -> Self {
+        Error::Engine(err)
```
Comment (BusyJay) [inline components/engine_traits/src/errors.rs:38]: This is too broad to be implemented. How about removing the trait and using `Error::Engine` directly?

Comment (koushiro) [reply components/engine_traits/src/errors.rs:38]: There may be dozens of implicit conversions about this in `engine_rocks`, I am not sure whether to remove this trait.

Comment (BusyJay) [reply components/engine_traits/src/errors.rs:38]: So terrible. 😓

File: components/engine_traits/src/errors.rs [hunk 2]
```diff
@@ -1,55 +1,41 @@
...
+pub enum Error {
+    // Engine uses plain string as the error.
+    #[error("Storage Engine {0}")]
+    Engine(String),
+    // FIXME: It should not know Region.
+    #[error(
+        "Key {} is out of [region {}] [{}, {})",
+        log_wrappers::Value::key(.0), .1, log_wrappers::Value::key(&.2), log_wrappers::Value::key(&.3)
+    )]
+    NotInRange(Vec<u8>, u64, Vec<u8>, Vec<u8>),
```
Comment (BusyJay) [inline components/engine_traits/src/errors.rs:19]: Can it be an anonymous struct?

File: components/engine_traits/src/lib.rs [hunk 3]
```diff
@@ -254,8 +254,6 @@
 #![feature(min_specialization)]
 #![recursion_limit = "200"]
```
Comment (BusyJay) [inline components/engine_traits/src/lib.rs:255]: This was set for `quick_error`, is it still necessary?

File: components/engine_traits/src/errors.rs (hunk 2 above)
Comment (koushiro) [reply components/engine_traits/src/errors.rs:19]: like this?

```
    #[error(
        "Key {} is out of [region {}] [{}, {})",
        log_wrappers::Value::key(&key), region_id, log_wrappers::Value::key(&start), log_wrappers::Value::key(&end)
    )]
    NotInRange {
        key: Vec<u8>, 
        region_id: u64, 
        start: Vec<u8>, 
        end: Vec<u8>,
    },
```

File: components/engine_traits/src/lib.rs (hunk 3 above)
Comment (koushiro) [reply components/engine_traits/src/lib.rs:255]: no

File: components/engine_traits/src/errors.rs (hunk 2 above)
Comment (BusyJay) [reply components/engine_traits/src/errors.rs:19]: Yes.

Comment (koushiro) [reply components/engine_traits/src/errors.rs]: done

**Review by BusyJay — APPROVED** [review]

**Review by nrc — APPROVED** [review]
awesome, thank you!

/lgtm

**Pull request merged**